Create scaffolding for Shape Detection API

litherum opened the pull request:
#### a132303fd6a0c4b8ee96b6cc986723477ffefb98
<pre>
Create scaffolding for Shape Detection API
<a href="https://bugs.webkit.org/show_bug.cgi?id=255215">https://bugs.webkit.org/show_bug.cgi?id=255215</a>
rdar://107821288

Reviewed by Mike Wyrzykowski.

The implementation of the Shape Detection API needs to run in the GPU process. However, in
WebKitLegacy, it can run in-process. Therefore, we&apos;re going to have to have 2 different
implementations: one that implements the functionality directly (that the GPU process and
WebKitLegacy will use) and another one that uses IPC to forward the calls across processes.

This patch implements this by creating a virtual interface for the API. This interface will
have 2 concrete implementations, and this patch stubs out one of them. The IPC implementation
will be implemented in another patch. This virtual interface is in
Source/WebCore/Modules/ShapeDetection/Interfaces, and the stubs for the first implementation
are in Source/WebCore/Modules/ShapeDetection/Implementation.

The original classes inside Source/WebCore/Modules/ShapeDetection, which are what the IDL
files target, delegate to this virtual interface. Each one of these classes has a Ref to a
&quot;backing&quot; object which is an implementation of this virtual interface.

No tests because there is no behavior change.

* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp:
(WebCore::BarcodeDetector::create):
(WebCore::BarcodeDetector::BarcodeDetector):
(WebCore::BarcodeDetector::getSupportedFormats):
(WebCore::BarcodeDetector::detect):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h:
(WebCore::BarcodeDetectorOptions::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h:
(WebCore::convertToBacking):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h:
(WebCore::DetectedBarcode::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.idl:
* Source/WebCore/Modules/ShapeDetection/DetectedFace.h:
(WebCore::DetectedFace::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedFace.idl:
* Source/WebCore/Modules/ShapeDetection/DetectedText.h:
(WebCore::DetectedText::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedText.idl:
* Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp:
(WebCore::FaceDetector::create):
(WebCore::FaceDetector::FaceDetector):
(WebCore::FaceDetector::detect):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.h:
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h:
(WebCore::FaceDetectorOptions::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp.
(WebCore::ShapeDetection::BarcodeDetectorImpl::BarcodeDetectorImpl):
(WebCore::ShapeDetection::BarcodeDetectorImpl::getSupportedFormats):
(WebCore::ShapeDetection::BarcodeDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp.
(WebCore::ShapeDetection::FaceDetectorImpl::FaceDetectorImpl):
(WebCore::ShapeDetection::FaceDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.cpp.
(WebCore::ShapeDetection::TextDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorOptionsInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeFormatInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedBarcodeInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedFaceInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedFace.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedTextInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedText.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/FaceDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/FaceDetectorOptionsInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/Landmark.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkTypeInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/LandmarkType.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/TextDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.h.
* Source/WebCore/Modules/ShapeDetection/Landmark.h:
(WebCore::Landmark::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Landmark.idl:
* Source/WebCore/Modules/ShapeDetection/LandmarkType.h:
(WebCore::convertToBacking):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Point2D.h:
(WebCore::Point2D::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Point2D.idl:
* Source/WebCore/Modules/ShapeDetection/TextDetector.cpp:
(WebCore::TextDetector::create):
(WebCore::TextDetector::TextDetector):
(WebCore::TextDetector::detect):
* Source/WebCore/Modules/ShapeDetection/TextDetector.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/262994@main">https://commits.webkit.org/262994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a1ca8b64275ec50cf5e1a167c671e3a93656ac0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3224 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3287 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4641 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/3569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3317 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2805 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3259 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/2903 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4453 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2875 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2790 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/2849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2927 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4197 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3303 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2874 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2876 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/794 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2871 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3141 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h:
```diff
@@ -25,13 +25,31 @@
 #pragma once
 
+#include "FaceDetectorOptionsInterface.h"
+#include <cstdint>
```
nit: This isn’t used by the changes you made to the header?  Is it due to unified build globbing change?

litherum replied:
Yep. It's for the `uint16_t`.

djg commented on Source/WebCore/Modules/ShapeDetection/DetectedBarcode.idl:
```diff
@@ -26,7 +26,8 @@
 [
-    EnabledBySetting=ShapeDetection
+    EnabledBySetting=ShapeDetection,
+    JSGenerateToJSObject
```
Why was it necessary to add `JSGenerateToJSObject`?

litherum replied:
because this object is returned from C++ to javascript (as opposed to just being delivered from javascript to C++), so it needs to have the toJSObject() function generated for it. This is the flag that tells our generators to do so.

djg reviewed (commented):
I’ve reviewed the C++ changes. I don’t feel confident to review the IDL changes and request someone else review those.

mwyrzykowski commented on Source/WebCore/Modules/ShapeDetection/Interfaces/FaceDetectorOptionsInterface.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <limits>
+
```
I guess for consistency and to avoid other unified build issues should we include cstdint here too?

litherum committed:
Create scaffolding for Shape Detection API
https://bugs.webkit.org/show_bug.cgi?id=255215
rdar://107821288

Reviewed by Mike Wyrzykowski.

The implementation of the Shape Detection API needs to run in the GPU process. However, in
WebKitLegacy, it can run in-process. Therefore, we're going to have to have 2 different
implementations: one that implements the functionality directly (that the GPU process and
WebKitLegacy will use) and another one that uses IPC to forward the calls across processes.

This patch implements this by creating a virtual interface for the API. This interface will
have 2 concrete implementations, and this patch stubs out one of them. The IPC implementation
will be implemented in another patch. This virtual interface is in
Source/WebCore/Modules/ShapeDetection/Interfaces, and the stubs for the first implementation
are in Source/WebCore/Modules/ShapeDetection/Implementation.

The original classes inside Source/WebCore/Modules/ShapeDetection, which are what the IDL
files target, delegate to this virtual interface. Each one of these classes has a Ref to a
"backing" object which is an implementation of this virtual interface.

No tests because there is no behavior change.

* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp:
(WebCore::BarcodeDetector::create):
(WebCore::BarcodeDetector::BarcodeDetector):
(WebCore::BarcodeDetector::getSupportedFormats):
(WebCore::BarcodeDetector::detect):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h:
(WebCore::BarcodeDetectorOptions::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h:
(WebCore::convertToBacking):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h:
(WebCore::DetectedBarcode::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.idl:
* Source/WebCore/Modules/ShapeDetection/DetectedFace.h:
(WebCore::DetectedFace::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedFace.idl:
* Source/WebCore/Modules/ShapeDetection/DetectedText.h:
(WebCore::DetectedText::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/DetectedText.idl:
* Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp:
(WebCore::FaceDetector::create):
(WebCore::FaceDetector::FaceDetector):
(WebCore::FaceDetector::detect):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.h:
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h:
(WebCore::FaceDetectorOptions::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp.
(WebCore::ShapeDetection::BarcodeDetectorImpl::BarcodeDetectorImpl):
(WebCore::ShapeDetection::BarcodeDetectorImpl::getSupportedFormats):
(WebCore::ShapeDetection::BarcodeDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp.
(WebCore::ShapeDetection::FaceDetectorImpl::FaceDetectorImpl):
(WebCore::ShapeDetection::FaceDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.mm: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.cpp.
(WebCore::ShapeDetection::TextDetectorImpl::detect):
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorOptionsInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeFormatInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedBarcodeInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedFaceInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedFace.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedTextInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/DetectedText.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/FaceDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetector.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/FaceDetectorOptionsInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/Landmark.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkTypeInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/LandmarkType.h.
* Source/WebCore/Modules/ShapeDetection/Interfaces/TextDetectorInterface.h: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.h.
* Source/WebCore/Modules/ShapeDetection/Landmark.h:
(WebCore::Landmark::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Landmark.idl:
* Source/WebCore/Modules/ShapeDetection/LandmarkType.h:
(WebCore::convertToBacking):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Point2D.h:
(WebCore::Point2D::convertToBacking const):
(WebCore::convertFromBacking):
* Source/WebCore/Modules/ShapeDetection/Point2D.idl:
* Source/WebCore/Modules/ShapeDetection/TextDetector.cpp:
(WebCore::TextDetector::create):
(WebCore::TextDetector::TextDetector):
(WebCore::TextDetector::detect):
* Source/WebCore/Modules/ShapeDetection/TextDetector.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/262994@main
